runtime/vm/vm_internal.h

```diff
@@ -553,8 +553,58 @@ void
 freeAllStructFFITypes(J9VMThread *currentThread, void *cifNode);
 #endif /* JAVA_SPEC_VERSION >= 16 */
 
+/* ------------------- jnimisc.cpp ----------------- */
+
+#if defined(J9VM_OPT_JAVA_OFFLOAD_SUPPORT)
+
+#define JAVA_OFFLOAD_SWITCH_ON_WITH_REASON_IF_LIMIT_EXCEEDED(currentThread, reason, length) \
+	do { \
+		if ((length) > J9_JNI_OFFLOAD_SWITCH_THRESHOLD) { \
+			javaOffloadSwitchOnWithReason(currentThread, reason); \
+		} \
+	} while (0)
+
+#define JAVA_OFFLOAD_SWITCH_OFF_WITH_REASON_IF_LIMIT_EXCEEDED(currentThread, reason, length) \
+	do { \
+		if ((length) > J9_JNI_OFFLOAD_SWITCH_THRESHOLD) { \
+			javaOffloadSwitchOffWithReason(currentThread, reason); \
+		} \
+	} while (0)
+
+#else /* defined(J9VM_OPT_JAVA_OFFLOAD_SUPPORT) */
+
+#define JAVA_OFFLOAD_SWITCH_ON_WITH_REASON_IF_LIMIT_EXCEEDED(currentThread, reason, length)
+#define JAVA_OFFLOAD_SWITCH_OFF_WITH_REASON_IF_LIMIT_EXCEEDED(currentThread, reason, length)
+
+#endif /* defined(J9VM_OPT_JAVA_OFFLOAD_SUPPORT) */
+
+/**
+ * @brief The helper method performs a memcpy from an object array to the native array.
+ *
+ * @param currentThread the thread performing the arraycopy
+ * @param arrayObject the object array to copy from
+ * @param isCopy a flag denoting the copy is completed upon return
+ * @param ensureMem32 a flag specific to the zOS/31bit
+ *
+ * @return the native array with the copied content
+ */
+void *
+memcpyFromObjectArray(J9VMThread *currentThread, j9object_t arrayObject, jboolean *isCopy, jboolean ensureMem32);
+
+/**
+ * @brief The helper method performs a memcpy to an object array from the native array.
+ *
+ * @param currentThread the thread performing the arraycopy
+ * @param arrayObject the object array to copy into
+ * @param elems the native array to copy to
+ * @param mode a code denoting whether to copy/release the native memory
+ * @param ensureMem32 a flag specific to the zOS/31bit
+*/
+void
+memcpyToObjectArray(J9VMThread *currentThread, j9object_t arrayObject, void *elems, jint mode, jboolean ensureMem32);
```

Comment:
Suggestions:
* this should identify the source file that implements these methods (like other declarations above)
* the function names are misleading - they don't operate on `Object[]` arrays
* improve the documentation
```c
/* ------------------- ArrayCopyHelpers.cpp ----------------- */

/**
 * @brief A helper method which creates a copy of a heap array in native memory.
 *
 * @param currentThread the thread performing the copy
 * @param arrayObject the heap array to copy from
 * @param isCopy NULL, or a pointer to a jboolean that will be set to JNI_TRUE if a copy is made
 * @param ensureMem32 a flag specific to 31-bit z/OS
 * @return the native array with the copied content
 */
void *
memcpyFromHeapArray(J9VMThread *currentThread, j9object_t arrayObject, jboolean *isCopy, jboolean ensureMem32);

/**
 * @brief A helper method which copies to a heap array from native memory,
 * then possibly releasing that native memory (to be paired with uses of
 * memcpyFromHeapArray).
 *
 * @param currentThread the thread performing the copy
 * @param arrayObject the heap array to copy into
 * @param elems the native memory to copy from
 * @param mode a code denoting whether to copy/release the native memory
 * @param ensureMem32 a flag specific to 31-bit z/OS
*/
void
memcpyToHeapArray(J9VMThread *currentThread, j9object_t arrayObject, void *elems, jint mode, jboolean ensureMem32);
```
I question the value of the `isCopy` parameter: callers can get the same information by checking if the pointer returned from `memcpyFromHeapArray()` is not `NULL`.

Reply:
> this should identify the source file that implements these methods (like other declarations above)
> the function names are misleading - they don't operate on Object[] arrays
> improve the documentation

Fixed as suggested above.

Reply:
> I question the value of the isCopy parameter: callers can get the same information by checking if the pointer returned from memcpyFromHeapArray() is not NULL.

`isCopy` (which is used everywhere in the JNI related code) is not intended to confirm whether the returned native pointer is a valid but whether the copy operation is done successfully. e.g. see https://github.com/eclipse-openj9/openj9/blob/master/runtime/tests/jni/jnierrors.c#L176, which is used in many JNI tests to validate the flag.

Frankly speaking, I didn't see the flag is literally used for something meaningful in our code but it doesn't hurt to keep it in there given the added code is shared with the existing JNI code.  If we remove it, should we remove it everywhere in our code?  e.g.
```
	/**
	 * Implementation of the JNI GetPrimitiveArrayCritical API.
	 * See the JNI spec for full details.
	 *
	 * @param vmThread		current J9VMThread (aka JNIEnv)
	 * @param array			a JNI reference to the primitive array
	 * @param isCopy		NULL, or a pointer to a jboolean that will set to JNI_TRUE if a copy is made <------
	 * @return 				a pointer to the array data; this may be a malloc'd copy, or a pointer into the heap
	 */
	virtual void* jniGetPrimitiveArrayCritical(J9VMThread* vmThread, jarray array, jboolean *isCopy) = 0;
```

Reply:
I wasn't proposing any changes to the any JNI interfaces.

I see two uses of `memcpyFromHeapArray()`:
* the call in `BytecodeInterpreter.hpp` doesn't care at all - the argument could simply be removed
* the call in `jnimisc.cpp` could be changed to
```c
		elems = memcpyFromHeapArray(currentThread, J9_JNI_UNWRAP_REFERENCE(array), ensureMem32);
		if ((NULL != elems) && (NULL != isCopy)) {
			*isCopy = JNI_TRUE;
		}
```

Reply:
It makes sense to me in this way and updated the code as proposed above.